File: src/com/suushiemaniac/lang/json/value/JsonType.java
package com.suushiemaniac.lang.json.value;

import com.suushiemaniac.lang.json.JSON;
import com.suushiemaniac.lang.json.exception.JsonNotIterableException;
import com.suushiemaniac.lang.json.exception.JsonValueTypeException;
import com.suushiemaniac.lang.json.util.StringUtils;

import java.util.*;
import java.util.stream.Stream;

public abstract class JsonType extends JSON {
	public static JSON fromNative(Object nativeObj) {
		if (nativeObj instanceof Boolean) {
			return new JsonBoolean((Boolean) nativeObj);
		} else if (nativeObj instanceof Float) {
			return new JsonFloat((Float) nativeObj);
		} else if (nativeObj instanceof Integer) {
			return new JsonInteger((Integer) nativeObj);
		} else if (nativeObj instanceof Void) {
			return new JsonNull();
		} else if (nativeObj instanceof String) {
			return new JsonString((String) nativeObj);
		}

		return null;
	}

	public class SingleIterator<T> implements Iterator<T> {
		private T toRead;
		private boolean hasRead;

		public SingleIterator(T element) {
			this.hasRead = false;
			this.toRead = element;
		}

		@Override
		public boolean hasNext() {
			return !hasRead;
		}

		@Override
		public T next() {
			hasRead = true;
			return toRead;
		}
	}

	protected JsonType(JSON parent) {
		super(parent);
	}

	public JSON get(JSON keyIndex) {
		throw new JsonNotIterableException("Can't get child elements from " + this.type());
	}

	public JSON get(String key) {
		throw new JsonNotIterableException("Can't get child elements from " + this.type());
	}

	public JSON get(int index) {
		throw new JsonNotIterableException("Can't get child elements from " + this.type());
	}

	public void set(JSON keyIndex, JSON value) {
		throw new JsonNotIterableException("Can't set type " + this.type() + " on key/index-based access");
	}

	public void set(String key, JSON value) {
		throw new JsonNotIterableException("Can't set type " + this.type() + " on key-based access");
	}

	public void set(int index, JSON value) {
		throw new JsonNotIterableException("Can't set type " + this.type() + " on index-based access");
	}

	public void set(JSON value) {
		throw new JsonNotIterableException("Invalid type for setting. Expected: " + this.type() + ", Found: " + value.type());
	}

	public void add(JSON value) {
		throw new JsonNotIterableException("Can't add to single value type " + this.type());
	}

	public void add(JSON keyIndex, JSON value) {
		throw new JsonNotIterableException("Can't add to single value type " + this.type());
	}

	public void add(String key, JSON value) {
		throw new JsonNotIterableException("Can't add to single value type " + this.type());
	}

	public void add(int index, JSON value) {
		throw new JsonNotIterableException("Can't add to single value type " + this.type());
	}

	public void remove(JSON value) {
		throw new JsonNotIterableException("Can't remove from single value type " + this.type());
	}

	public void remove(int index) {
		throw new JsonNotIterableException("Can't remove from single value type " + this.type());
	}

	public void remove(String key) {
		throw new JsonNotIterableException("Can't remove from single value type " + this.type());
	}

	public JSON keyIndexOf(JSON content) {
		throw new JsonNotIterableException("Single value types don't have any keys/indices");
	}

	public void clear() {
		throw new JsonNotIterableException("Can't clear single value types");
	}

	public int size() {
		return 1;
	}

	public int deepSize() {
		return 1;
	}

	public String toFormatString() {
		return StringUtils.copy("\t", this.hierarchy()) + this.toString();
	}

	public String toXMLString() {
		return this.toString();
	}

	public boolean booleanValue() {
		throw new JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Boolean");
	}

	public int intValue() {
		throw new JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Integer");
	}

	public float floatValue() {
		throw new JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Float");
	}

	public String stringValue() {
		throw new JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: String");
	}

	public Object nullValue() {
		throw new JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Null");
	}

	public Collection<JSON> collect() {
		return Collections.singleton(this);
	}

	public Set<JSON> keySet() {
		return Collections.emptySet();
	}

	@Override
	public Stream<JSON> stream() {
		return Stream.of(this);
	}

	@Override
	public Map<String, Object> nativeMap() {
		throw new JsonNotIterableException();
	}

	@Override
	public List<Object> nativeList() {
		return Collections.singletonList(this);
	}

	@Override
	public Set<Object> nativeSet() {
		return Collections.singleton(this);
	}

	@Override
	public Iterator<JSON> iterator() {
		return new SingleIterator<>(this);
	}
}

File: src/main/kotlin/com/suushiemaniac/lang/json/value/JsonType.kt
package com.suushiemaniac.lang.json.value

import com.suushiemaniac.lang.json.JSON
import com.suushiemaniac.lang.json.exception.JsonNotIterableException
import com.suushiemaniac.lang.json.exception.JsonValueTypeException
import com.suushiemaniac.lang.json.util.StringUtils.decodeUnicode

abstract class JsonType protected constructor(parent: JSON?) : JSON(parent) {
    inner class SingleIterator<T>(private val toRead: T) : Iterator<T> {
        private var hasRead: Boolean = false

        override fun hasNext(): Boolean {
            return !hasRead
        }

        override fun next(): T {
            hasRead = true
            return toRead
        }
    }

    override fun get(keyIndex: JSON): JSON? {
        throw JsonNotIterableException("Can't get child elements from " + this.type())
    }

    override fun get(key: String): JSON? {
        throw JsonNotIterableException("Can't get child elements from " + this.type())
    }

    override fun get(index: Int): JSON? {
        throw JsonNotIterableException("Can't get child elements from " + this.type())
    }

    override fun set(keyIndex: JSON, value: JSON) {
        throw JsonNotIterableException("Can't set type " + this.type() + " on key/index-based access")
    }

    override fun set(key: String, value: JSON) {
        throw JsonNotIterableException("Can't set type " + this.type() + " on key-based access")
    }

    override fun set(index: Int, value: JSON) {
        throw JsonNotIterableException("Can't set type " + this.type() + " on index-based access")
    }

    override fun set(value: JSON) {
        throw JsonNotIterableException("Invalid type for setting. Expected: " + this.type() + ", Found: " + value.type())
    }

    override fun add(value: JSON) {
        throw JsonNotIterableException("Can't add to single value type " + this.type())
    }

    override fun add(keyIndex: JSON, value: JSON) {
        throw JsonNotIterableException("Can't add to single value type " + this.type())
    }

    override fun add(key: String, value: JSON) {
        throw JsonNotIterableException("Can't add to single value type " + this.type())
    }

    override fun add(index: Int, value: JSON) {
        throw JsonNotIterableException("Can't add to single value type " + this.type())
    }

    override fun remove(value: JSON) {
        throw JsonNotIterableException("Can't remove from single value type " + this.type())
    }

    override fun remove(index: Int) {
        throw JsonNotIterableException("Can't remove from single value type " + this.type())
    }

    override fun remove(key: String) {
        throw JsonNotIterableException("Can't remove from single value type " + this.type())
    }

    override fun keyIndexOf(content: JSON): JSON {
        throw JsonNotIterableException("Single value types don't have any keys/indices")
    }

    override fun clear() {
        throw JsonNotIterableException("Can't clear single value types")
    }

    override fun size(): Int {
        return 1
    }

    override fun depth(): Int {
        return 0
    }

    override fun deepSize(): Int {
        return 1
    }

    override fun toFormatString(): String {
        return "\t".repeat(this.hierarchy()) + this.toString()
    }

    override fun toXMLString(): String {
        return this.toString()
    }

    override fun booleanValue(): Boolean {
        throw JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Boolean")
    }

    override fun intValue(): Int {
        throw JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Integer")
    }

    override fun floatValue(): Float {
        throw JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Float")
    }

    override fun stringValue(): String {
        throw JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: String")
    }

    override fun decodedStringValue(): String {
        return this.stringValue().decodeUnicode()
    }

    override fun nullValue(): Any {
        throw JsonValueTypeException("Accessing by incompatible value. Expected: " + this.type() + ", Found: Null")
    }

    override fun collect(): Collection<JSON> {
        return setOf<JSON>(this)
    }

    override fun keySet(): Set<JSON> {
        return emptySet()
    }

    override fun nativeKeySet(): Set<String> {
        return emptySet()
    }

    override fun nativeMap(): Map<String, Any> {
        throw JsonNotIterableException()
    }

    override fun <T> nativeMap(mapping: (JSON) -> T): Map<String, T> {
        throw JsonNotIterableException()
    }

    override fun nativeList(): List<Any> {
        return listOf<Any>(this)
    }

    override fun <T> nativeList(mapping: (JSON) -> T): List<T> {
        return listOf(mapping(this))
    }

    override fun nativeSet(): Set<Any> {
        return setOf<Any>(this)
    }

    override fun <T> nativeSet(mapping: (JSON) -> T): Set<T> {
        return setOf(mapping(this))
    }

    override fun iterator(): Iterator<JSON> {
        return SingleIterator<JSON>(this)
    }

    companion object {
        fun fromNative(nativeObj: Any?): JSON? {
            return when (nativeObj) {
                is Boolean -> JsonBoolean(nativeObj)
                is Float -> JsonFloat(nativeObj)
                is Double -> JsonFloat(nativeObj)
                is Int -> JsonInteger(nativeObj)
                is Long -> JsonInteger(nativeObj)
                is Short -> JsonInteger(nativeObj)
                is Byte -> JsonInteger(nativeObj)
                is String -> JsonString(nativeObj)
                null -> JsonNull.INST
                else -> null
            }
        }
    }
}
